### ISSUE TITLE
Stratified K Fold and Dataset

### ISSUE THREAD

**Issue by crisbodnar** [rusty1s/pytorch_geometric#1172]
## 🐛 Bug

Hi there! The Pytorch geometric Dataset object used to work nicely with scikit-learn's StratifiedKFold. See an example below:
```
 kf2 = StratifiedKFold(n_splits=9, shuffle=False)
 for train_idxs, val_idxs in kf2.split(dataset, dataset.data.y):
     train_dataset = dataset[torch.LongTensor(train_idx)]
```

In version 1.3.2 the dataset slicing used to also affect dataset.data.y. However, slicing now seems to leave this unchanged. So if one wants to further split train_dataset using StratifiedKFold, this is not working anymore because len(train_dataset) and len(train_dataset.data.y) won't match. 

## To Reproduce

1. Load Proteins dataset from PyTorch geometric
1. Check data len(dataset[indices]) is different from len(dataset[indices].data.y). 

Is this the new expected behaviour and is there any elegant workaround for this? 

I think the older behaviour from version 1.3.2 was more natural. 

## Environment

* OS: MacOS
* Python version: 3.6
* PyTorch version: 1.5
* CUDA/cuDNN version: cpu

**Comment by rusty1s** [comment]
Hi! I partial agree with you that this was more natural back then. Part of the reason I changed that was that each time you sliced the dataset, you re-created it in memory (which is quite unnecessary). Let me think of a way to get the best of both worlds :)

To prevent this behaviour, you can always do something like:
```
y = torch.cat([data.y for data in dataset])
```

**Comment by crisbodnar** [comment]
Thanks a lot for the quick reply! That makes sense. 

Perhaps the old behaviour could be supplied through some utils function or a dataset.subset(indices) method in the dataset object?

**Comment by fclairec** [comment]
Hi, thanks for this. 
I am now wondering how I can replace the y labels in the train and validation set, such that they meet the length of the dataset. My guess would have been like this:
![image](https://user-images.githubusercontent.com/59439696/81553633-baac2a00-9385-11ea-9e1f-7c1b38f8d50b.png)

After line 63 the lengths of the validation dataset are equal. However it seems that in line 66 my val_dataset.data.y is affected and further reduced when I reset the train_dataset.data.y ...

Do you see a quick fix for this? I don't fully understand what is happening here.

**Comment by rusty1s** [comment]
You should never modify `dataset.data` directly. If you really want to modify `dataset.data`, it is best to follow the `InMemoryDataset` guidelines:
```python
data_list = [data for data in dataset[train_idx]]
dataset.data, dataset.slices = self.collate(data_list)
```

**Comment by fclairec** [comment]
Thanks for your quick reply. That works better!

**Comment by rusty1s** [comment]
I just added a `dataset.copy()` method that converts sliced datasets back to a contiguous layout:
```
from torch_geometric.datasets import TUDataset

dataset = TUDataset('/tmp/TUDataset', 'MUTAG')
print(len(dataset))
>>> 180
print(dataset.data.y.size())
>>> torch.Size([188])

dataset = dataset[:40]
print(len(dataset))
>>> 40
print(dataset.data.y.size())
>>> torch.Size([188])

dataset = dataset.copy()
print(len(dataset))
>>> 40
print(dataset.data.y.size())
>>> torch.Size([40])
```

**Comment by fclairec** [comment]
Maybe for the completeness the new method could be extended to include the preparation of a subset based on the indices prepared before hand. (e.g. the initial post was generating indices with StratifiedKFold)

`
    def copy(self, idx=None):
        if idx is not None:
            data_list = [self.get(i) for i in idx]
        else:
            data_list = [self.get(i) for i in range(len(self))]
        dataset = copy.copy(self)
        dataset.__indices__ = None
        dataset.data, dataset.slices = self.collate(data_list)
        return dataset`

**Comment by rusty1s** [comment]
Done! :)

**Comment by tunglm2203** [comment]
I used [copy()](https://github.com/rusty1s/pytorch_geometric/blob/4b531e5168555c5c2600cc616a21aa050306efb4/torch_geometric/data/in_memory_dataset.py#L138) function to create another dataset from a dataset given indices, as in the newest commit, the copy() is defined as follows:

```
def copy(self, idx=None):
        if idx is None:
            data_list = [self.get(i) for i in range(len(self))]
        else:
            data_list = [self.get(i) for i in idx]
        dataset = copy.copy(self)
        dataset.__indices__ = None
        dataset.__data_list == data_list
        dataset.data, dataset.slices = self.collate(data_list)
        return dataset
```
I check the [get()](https://github.com/rusty1s/pytorch_geometric/blob/4b531e5168555c5c2600cc616a21aa050306efb4/torch_geometric/data/in_memory_dataset.py#L69) function calls to `self.__data_list__`, so the line `dataset.__data_list == data_list` should be modified to `dataset.__data_list__ == data_list` (with double underscores after `__data_list`), isn't it?
